Various changes from pybinding

jagerman opened the pull request:
- Change `listen` return to `void`: it only ever throws, or returns true, so convert it to void with a void promise so now it either throws on failure, or doesn't throw on success.
- Moved the direction methods up to `connection_interface` so that a public caller (like pybind) can get them.
- Removed a dubious-looking double-enable_shared_from_this inheritance from Connection

jagerman committed:
Change `listen` return to `void`

It never returns false: it only ever throws, or returns true, so convert
it to void with a void promise so now it either throws on failure, or
doesn't throw on success.

jagerman committed:
Move direction methods to connection_interface

So that it is accessible when you have the generic one, e.g. from
pybind.

jagerman committed:
Remove dubious double-enable_shared_from_this

The one in connection_interface is used, but the second one in
Connection doesn't appear to be, and seems potentially dangerous: we
have two different shared_pointers inside an Connection instance and it
isn't clear that they are aware of each other.

Removing it affected nothing else, so seems fine to drop.

jagerman committed:
Make `send(bstring)` less dangerous

Currently we don't allow calling `send` with a std::string or ustring,
but a bstring slips through the cracks via the base method
implementation: calling `send(my_bstring)` ends up calling effectively
`send(bstring_view{my_bstring})`, but this ends up sending garbage
because that `my_bstring` likely doesn't live long enough.

This rearranges the virtual method to be protected so that bstring gets
the same protection as std::string and ustring, avoiding an error-prone
case.

(I encountered this in innocuous-looking code that was doing exactly
this, and ending up sending garbage on the wire).

*With* this change, `send(my_bstring)` no longer compiles, forcing you
to change it to `send(std::move(my_bstring))` or else doing something
with a keep-alive, as appropriate.

jagerman committed:
Don't fail for sending an empty string

There's nothing specifically *wrong* with sending an empty string: it
just doesn't do anything (just like you can print an empty string to a
file or stdout and have it do nothing).

This was throwing (unexpected) exceptions when I casually sent some data
that happened to be empty, and it felt weird to have to make the
application do a `if (!empty)` everywhere it sends something.

jagerman committed:
Format

jagerman committed:
Add stream buffer ownership transfer test case

I wrote this up while investigating a crash on the Python wrapper side.
This passes just fine here (and so it's a python wrapper issue), but
having written it it seems a worthwhile test to have.

(We have other tests that explicitly rely on these things, but not one
that definitely tests just this one issue).

dr7ana commented on src/stream.cpp:
```diff
@@ -161,7 +161,7 @@ namespace oxen::quic
     void Stream::send_impl(bstring_view data, std::shared_ptr<void> keep_alive)
     {
         if (data.empty())
-            throw std::invalid_argument{"Cannot send empty byte string"};
```
I honestly kinda like that this throws on an empty send? It immediately tells us that something is wrong in whatever application logic is trying to feed data to libquic

Though I've been off in refactor land for a long time, def gonna pick your brain off band on this

jagerman replied:
> I honestly kinda like that this throws on an empty send? It immediately tells us that something is wrong in whatever application logic is trying to feed data to libquic

The thing is, it only *might* be a bug in application logic, but by throwing here we're forcing it to be an application bug.  It's kind of like `std::cout << "";` -- it does nothing, but isn't necessarily a bug.

jagerman replied:
Case in point: my echo-server broke without this because it reads a line of input, then sends it, and broke when I hit enter without typing anything.  It feels weird for that to be an invalid use of `send()` and for me to have to make sure that anywhere I call send() I've already checked for non-emptiness.

dr7ana replied:
Shit. Valid.

jagerman committed:
Hold dgram_interface in a shared_ptr

This aligns it with how Stream works, and makes it much easier to use
ith pybind (where pybind wants to be able to share ownership via a
shared_ptr holder).

dr7ana committed:
linter, ci

jagerman commented on include/quic/datagram.hpp:
```diff
@@ -47,7 +47,7 @@ namespace oxen::quic
 
     // IO callbacks
-    using dgram_data_callback = std::function<void(dgram_interface&, bstring)>;
+    using dgram_data_callback = std::function<void(std::shared_ptr<dgram_interface>, bstring)>;
```
We do not want to pass a shared_ptr here (or for the stream data): it will incur the overhead of instantiating an additional shared_ptr instance (which requires an atomic reference count increment on construction and decrement on destruction) on every data callback, even if the caller does not care about or need a shared ptr.  By passing it as a reference we avoid that overhead and only incur if if the caller needs it (by calling `shared_from_this()` on the reference).

The relatively infrequent calls (stream open, etc.) could be passed by shared_ptr, but I'm not sure it's worthwhile to be inconsistent (references for data callbacks, shared_ptr for other things).

jagerman committed:
Fix race condition in stream()

The weak pointer could away between the call to `expired()` and the
subsequent `lock()`.  `lock()`ing first, then checking that lock
returned non-empty avoids the race.

jagerman committed:
Update oxen-encoding (for bstring support)

jagerman committed:
Fail for bad input on request length parser

Currently the logic here continually accumulates a buffer until a `:` is
found, but someone could abuse that by sending a very large stream of
data that never contains a `:` to run the receiver out of memory.  This
change puts a maximum on the length parser to mitigate this.

(There was also a bug here that was checking `.back()` instead of
`npos`)

jagerman committed:
Convert btstream to use more bstring and bstring_views

There were some unnecessary copies in here, but also using a bstring
matches much better what we're actually sending on the wire,
particularly when interacting with Python.  The internals now use
bstring/bstring_views and the public-facing APIs now take the body as
either a regular string_view or a bstring_view.

jagerman committed:
Fix issues with multiple sequential bt requests

Adds a test that blasts 50 requests at once, which triggered some
failures (fixed here):

- the request id lower_bound lookup had a reversed condition and, as a
  result, didn't properly find requests when there were more than one in
  the container.

- The `buf` we build up to pass to handle_input had `current_len` added
  to it, but wasn't taking into account that we might already have some
  of the request loaded into `buf` from a previous request (and so we
  could end up losing some of the data needed for processing the next
  request).

dr7ana committed:
Address:set_port

jagerman committed:
Add some tests of big request success and bigger request failures

jagerman committed:
Fix class/struct mismatch

jagerman committed:
Error handling cleanup

This refactors the handling of error codes to address a few
shortcomings and problems:

- error codes were kind of scattered all over, but in terms of numeric
  value, and in terms of where they are defined.  This concentrates them
  both numerically and in the code (now in error.hpp).
- error codes are now all values above 777'000'000, which makes them
  easier to read out of logs (e.g. 777000100: 777 means a libquic error,
  and 100 indicates stream exception).  Previously we were using values
  under 2<<62, but those seem pointlessly huge (and aren't trivially
  easy to decipher out of a log).
- the error code for an expiring connection was invalid (2<<62 + 1,
  which is larger than the max allowed QUIC integer), and so when a
  connection timed out an ngtcp2 assert failed and aborted the program.
- fixed some weirdness in stream.close() that took an io_error and sent
  an *ngtcp2* error code, but that wasn't right because those are
  internal, negative codes: we need to always send a uint64_t
  application error code.  This changes close to simply take an error
  code.
- added a custom exception class (`application_stream_error`) that
  application code can throw to fire a custom error code down the
  stream (this solves a TODO in the code).
- redid quic_strerror to return string that can also display a custom
  error code, if used.
- removed the `error::` enum because it felt kind of wrong to be
  static_cast'ing uint64_t values into the error:: enum for enum values
  that it didn't actually hold.  (This is valid C++, but feels a bit
  dirty).

jagerman committed:
Simplify make_command/make_response

They were doing a bit too much: opt:: values got passed through, as well
as rid incrementing and exception handling, plus the actual encoding.
This renames them to `encode_command()`/`encode_response()` and has them
only do the encoding part, leaving the rest to the existing caller which
splits up the logic involved fairly cleanly.

jagerman committed:
Misc cleanups

- change stream count variables to uint64_t; having them as int's
  introduces a bunch of pain when interacting with the ngtcp2
  input/outputs that are uint64_t.
- remove redundant BTRequestStream destructor (it was doing part of what
  post-destructor destruction has to do anyway).
- address a bunch of clang-tidy warnings about using reserved
  double-underscore prefixes in variables

jagerman committed:
Debug-log connection destruction in debug builds

jagerman committed:
Fix idle close crash

By not causing `set_closing()` for the ERR_IDLE_CLOSE case we weren't
safe against repeated closes, and so if a connection close happened
during stream destruction after an idle close (as lokinet was doing)
we'd double-close and crash.

jagerman committed:
Make public close_connection interface thread-safe

Lokinet was (innocently) calling this from another thread, but that can
and will break things.

Also fixes a potential crash with `close_conns` that could suffer from
iterator invalidation because of close_connection() changing `conns`
while it iterates.

jagerman committed:
Defer libquic object destructions to the event loop

This adds a `.make_shared<T>` to Network and Endpoint that works almost
like std::make_shared<T>, except that it ensures that the destructor
will be invoked in the Network's event loop, which is not necessarily
the thread that releases the last shared_ptr.

This is implemented by a two-pass custom deleter: when the last
shared_ptr is released, rather than immediately triggering destruction
(as in a standard shared_ptr) we instead queue the raw pointer
destruction onto the event loop thread to perform the destruction.

This commit also makes the stream, btstream, and datagram constructors
protected, with Network as a friend class, so that they *have* to go
through net.make_shared (because std::make_shared will produce a
shared_ptr that destructs in the wrong place).

jagerman committed:
Allow stream_construct_cb to return nullptr

If it returns nullptr we construct a default stream.  This makes the
callback a little more flexible: it doesn't have to worry about how a
default stream should be constructed, but can only worry about special
streams.

dr7ana committed:
expose remote pubkey from connection interface

dr7ana committed:
expose get_stream from connection level

dr7ana committed:
view remote_key method tweak

dr7ana committed:
templated stream getter is better

jagerman commented on include/quic/connection.hpp:
```diff
@@ -90,6 +91,12 @@ namespace oxen::quic
+        template <typename StreamT = Stream, std::enable_if_t<std::is_base_of_v<Stream, StreamT>, int> = 0>
+        std::shared_ptr<StreamT> get_stream(int64_t id)
+        {
+            return std::static_pointer_cast<StreamT>(get_stream_impl(id));
```
This needs to be a dynamic_pointer_cast if you can't guarantee that the pointer is actually a StreamT

jagerman committed:
Fix `message` move/copy constructors

Moving sometimes, and copying always, would leave you with a `data`
member with different memory from before, but all the string_view's
pointing into the *old* data member.

This adds explicit copy/move ctors/assignments that properly update all
the string_views to point into the new member.

jagerman committed:
Fix segv in check_timeout

`f` never got updated to point to the new first element after erasing.

jagerman committed:
(squash me) message: nicer solution to copy/move issues

Updating the string_views was a bit hairy when, essentially, what we
need are offset/sizes of the substr within `data`.  The changes the
implementation to just store offset/size pairs, rather than
string_views, which simplifies things a lot as we can now just use the
default copy/move constructors again.

dr7ana commented on include/quic/btstream.hpp:
```diff
@@ -29,23 +29,22 @@ namespace oxen::quic
+        // We keep the locations of variables fields as relative positions inside `data` *rather*
+        // than using std::string_view members because the string_views are more difficult to
+        // maintain when the object gets moved or copied.
+        using substr_location = std::pair<std::ptrdiff_t, std::size_t>;
```
This would be a really cool idea for a PR to oxen-encoding, basically extracting a map or some referential structure indicating the indexes of each data field in the bt-dict or list

dr7ana commented on tests/002-send-receive.cpp:
```diff
@@ -269,6 +351,206 @@ namespace oxen::quic::test
+
+            CHECK(done.wait_for(5s) == std::future_status::ready);
+            std::lock_guard lock{mut};
+            CHECK(good_responses == num_requests);
```
Okay so I dug into why this is failing literally all the CI builds. I separated the test cases to look into the one that was failing specifically, pushed it to [my branch](https://github.com/dr7ana/libquic/blob/f20358f345e463c773c41452972a7a7764ca0f6f/tests/002-send-receive.cpp#L432) to keep testnet tracking the pr void-listen

[This is the offending line](https://github.com/dr7ana/libquic/blob/f20358f345e463c773c41452972a7a7764ca0f6f/src/connection.cpp#L910) that slows down the test to the point of failure, and downgrading it's log-level to trace makes the test work as you wrote it.

Though what's interesting is if you leave that line as a debug log, put [this line](https://github.com/dr7ana/libquic/blob/f20358f345e463c773c41452972a7a7764ca0f6f/tests/002-send-receive.cpp#L513) instead of the `done.wait_for(5s)` call, and run the test suit with debug or trace logging, it loops forever (both with and without the mutexes)

jagerman committed:
Fix wrong type in enable_if for operator=

jagerman committed:
Fix crash in register_command

If not in the logic thread, the deferred call here was keeping invalid
references.

jagerman committed:
Fix from_chars not checking that we consumed the whole string

As written it would accept something like "123abc:" as length 123 and
just ignore the "abc", but we should be failing in such a case.

dr7ana committed:
Stream construction, message timeouts?

- We need the stream id to be passed in stream construction to specify btreq streams
- message timeout vs error needs to be re-granulared

dr7ana committed:
testing something

dr7ana committed:
stream count method

dr7ana committed:
configurable keep-alive timeout

dr7ana committed:
oxenc vbump

jagerman committed:
Make sure there is no garbage after bt request stream data

jagerman committed:
Disable over-logging of big request tests

The test here produces 500k lines of debug logging when at debug log
level, so force it up to (at least) info for the duration of the test.

dr7ana commented on src/btstream.cpp:
```diff
@@ -25,6 +25,8 @@ namespace oxen::quic
 
         req_body = get_location(data, btlc.consume_string_view());
+
+        btlc.finish();
```
I dunno about this change i really need to think about it.

@tewinget thoughts?
